### PR TITLE
Bugfix/hashmap int overflow

### LINKED ISSUE #409: Type conversion (maybe corrupting data)?
Hi 

I am trying to update my R package sundialr which is an interface to some of the solvers in SUNDIALS. The CRAN checks for my package give -Woverflow warnings. I discussed this issue in r-package-devel forum where another user thought I should bring it to the attention of the library since it may be corrupting data.

See the entire issue here - 

https://stat.ethz.ch/pipermail/r-package-devel/2024q1/010431.html

I am not able to submit to CRAN due to this error.

Specifically, these are the warnings I am getting

>Flavor: r-devel-windows-x86_64
>Check: whether package can be installed, Result: WARNING
>  Found the following significant warnings:
>    ./sundials/sundials/sundials_hashmap.h:26:48: warning: conversion from
>'long long unsigned int' to 'long unsigned int' changes value from
>'14695981039346656037' to '2216829733' [-Woverflow]
>    ./sundials/sundials/sundials_hashmap.h:27:48: warning: conversion from
>'long long unsigned int' to 'long unsigned int' changes value from
>'1099511628211' to '435' [-Woverflow]
>    sundials/sundials/sundials_hashmap.h:26:48: warning: conversion from
>'long long unsigned int' to 'long unsigned int' changes value from
>'14695981039346656037' to '2216829733' [-Woverflow]
>    sundials/sundials/sundials_hashmap.h:27:48: warning: conversion from
>'long long unsigned int' to 'long unsigned int' changes value from
>'1099511628211' to '435' [-Woverflow]
>    sundials/sundials/sundials_profiler.c:71:24: warning: function
>declaration isn't a prototype [-Wstrict-prototypes]
>  See 'd:/RCompile/CRANincoming/R-devel/sundialr.Rcheck/00install.out' for
>details.
>  Used C++ compiler: 'g++.exe (GCC) 12.3.0'

and 

>Flavor: r-devel-linux-x86_64-debian-gcc
>Check: whether package can be installed, Result: WARNING
>  Found the following significant warnings:
>    sundials/sundials/sundials_profiler.c:71:41: warning: a function
>declaration without a prototype is deprecated in all versions of C
>[-Wstrict-prototypes]
>  See '/srv/hornik/tmp/CRAN/sundialr.Rcheck/00install.out' for details.
>  Used C++ compiler: 'Debian clang version 17.0.6 (5)'

Any update on this would be highly appreciated!

### PR DESCRIPTION
**Pull request by balos1**
Fixes #409 and #249.

### CODE REVIEW

**Review by drreynolds — APPROVED** [review]

File: doc/arkode/guide/source/Introduction.rst [hunk 1]
```diff
@@ -313,6 +313,10 @@ rely on these are recommended to transition to the corresponding :c:type:`SUNMat
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```
Comment (gardner48) [inline doc/arkode/guide/source/Introduction.rst:317]:
```suggestion
`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```

File: doc/cvode/guide/source/Introduction.rst [hunk 2]
```diff
@@ -283,6 +283,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```
Comment (gardner48) [inline doc/cvode/guide/source/Introduction.rst:287]:
```suggestion
`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```

File: doc/cvodes/guide/source/Introduction.rst [hunk 3]
```diff
@@ -283,6 +283,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```
Comment (gardner48) [inline doc/cvodes/guide/source/Introduction.rst:287]:
```suggestion
`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```

File: doc/ida/guide/source/Introduction.rst [hunk 4]
```diff
@@ -244,6 +244,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```
Comment (gardner48) [inline doc/ida/guide/source/Introduction.rst:248]:
```suggestion
`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```

File: doc/idas/guide/source/Introduction.rst [hunk 5]
```diff
@@ -258,6 +258,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```
Comment (gardner48) [inline doc/idas/guide/source/Introduction.rst:262]:
```suggestion
`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```

File: doc/kinsol/guide/source/Introduction.rst [hunk 6]
```diff
@@ -239,6 +239,10 @@ corresponding :c:type:`SUNMatrix` and :c:type:`SUNLinearSolver` modules:
 Fixed `GitHub Issue #329 <https://github.com/LLNL/sundials/issues/329>`_ so
 that C++20 aggregate initialization can be used.
 
+Fixed integer overflow in the internal SUNDIALS hashmap. This resolves
+`GitHub Issue #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```
Comment (gardner48) [inline doc/kinsol/guide/source/Introduction.rst:243]:
```suggestion
`GitHub Issues #409 <https://github.com/LLNL/sundials/issues/409>`_ and
```

File: src/sundials/sundials_hashmap.h [hunk 7]
```diff
@@ -20,14 +20,15 @@
...
 #include <stdlib.h>
 #include <string.h>
 
 #include "sundials/sundials_errors.h"
 #include "sundials/sundials_types.h"
 
-static const unsigned long HASH_PRIME        = 14695981039346656037U;
-static const unsigned long HASH_OFFSET_BASIS = 1099511628211U;
+static const uint64_t HASH_PRIME        = 14695981039346656037U;
+static const uint64_t HASH_OFFSET_BASIS = 1099511628211U;
```
[the remark this answers is not in the document]
Comment (balos1) [reply src/sundials/sundials_hashmap.h:31]: I don't think so, and it seems like the platform dependence of `unsigned long long` could only be a downside in this context.

File: src/sundials/sundials_hashmap.h [hunk 8]
```diff
@@ -36,9 +37,9 @@ static const unsigned long HASH_OFFSET_BASIS = 1099511628211U;
   This is a 64-bit implementation of the 'a' modification of the
   Fowler–Noll–Vo hash (i.e., FNV1-a).
  */
-static unsigned long fnv1a_hash(const char* str)
+static uint64_t fnv1a_hash(const char* str)
 {
-  unsigned long hash = HASH_OFFSET_BASIS;
+  uint64_t hash = HASH_OFFSET_BASIS;
```
[the remark this answers is not in the document]
Comment (balos1) [reply src/sundials/sundials_hashmap.h:42]: It does not cast the return value of `fnv1a_hash` to an `int`. It casts the result of `fnv1a_hash` modulo `max_size`, which should always fit in an `int`.

**Review by gardner48 — APPROVED** [review]

**Pull request merged**